gneiss/util.py

```diff
@@ -0,0 +1,172 @@
+
+
+def match_tips(table, tree, intersect=False):
+    """ Returns the contingency table and tree with matched tips.
+
+    Sorts the columns of the contingency table to match the tips in
+    the tree.  The ordering of the tips is in post-traversal order.
+
+    If the tree is multi-furcating, then the tree is reduced to a
+    bifurcating tree by randomly inserting internal nodes.
+
+    Parameters
+    ----------
+    table : pd.DataFrame
+        Contingency table where samples correspond to rows and
+        features correspond to columns.
+    tree : skbio.TreeNode
+        Tree object where the leafs correspond to the features.
+    intersect : bool, optional
+        Specifies if only the intersection of samples in the
+        contingency table and the tree will returned.
+        By default, this is False.
+
+    Returns
+    -------
+    pd.DataFrame :
+        Subset of the original contingency table with the common features.
+    skbio.TreeNode :
+        Sub-tree with the common features.
+
+    Raises
+    ------
+    ValueError:
+        Raised if `table` and `tree` have incompatible sizes.
+
+    See Also
+    --------
+    skbio.TreeNode.bifurcate
+    skbio.TreeNode.tips
+    """
+    tips = [x.name for x in tree.tips()]
+    common_tips = list(set(tips) & set(table.columns))
+
+    if intersect:
+        _table = table.loc[:, common_tips]
+        _tree = tree.shear(names=common_tips)
+    else:
+        if len(tips) != len(table.columns):
+            raise ValueError("`table` and `tree` have incompatible sizes, "
+                             "`table` has %d columns, `tree` has %d tips.  "
+                             "Consider setting `intersect=True`." %
+                             (len(table.columns), len(tips)))
+
+        _table = table
+        _tree = tree
+
+    _tree.bifurcate()
+    _tree.prune()
+    sorted_features = [n.name for n in _tree.tips()]
+    _table = _table.reindex_axis(sorted_features, axis=1)
+    return _table, _tree
+
+
+def rename_internal_nodes(tree, names=None):
+    """ Names the internal according to level ordering.
+
+    The tree will be traversed in level order (i.e. top-down, left to right).
+    If `names` is not specified, the node with the smallest label (y0)
+    will be located at the root of the tree, and the node with the largest
+    label will be located at bottom right corner of the tree.
+
+    Parameters
+    ----------
+    tree : skbio.TreeNode
+        Tree object where the leafs correspond to the features.
+    names : list, optional
+        List of labels to rename the tip names.  It is assumed that the
+        names are listed in level ordering, and the length of the list
+        is at least as long as the number of internal nodes.
+
+    Returns
+    -------
+    skbio.TreeNode
+       Tree with renamed internal nodes.
+
+    ValueError:
+        Raised if `tree` and `name` have incompatible sizes.
+    """
+    _tree = tree.copy()
```

Comment:
The copy isn't _strictly_ necessary.  But it is certainly useful.  Just added an `inplace` parameter here.  

It'll be tricky to have this functionality in the other modules, mainly because pandas and skbio don't make filter and `shear` operations in place (as far as I'm aware).  If you think this is necessary, we can create an issue for this, so that this PR can proceed.